src/parser/parser.cc

```diff
@@ -0,0 +1,1103 @@
+    }
+
+    for (auto type_def : defs.types) {
+      types.Set(type_def->header, type_def);
+    }
+
+    return IRModule(funcs, types);
+  }
+
+  SemVer ParseSemVer() {
+    // Consume(TokenType::Unknown);
+    return SemVer{.major = 0, .minor = 0, .patch = 0};
+  }
+
+  Definitions ParseDefinitions() {
+    Definitions defs;
+
+    while (true) {
+     auto next = Peek();
+     switch (next->token_type) {
+        case TokenType::Defn: {
+          Consume(TokenType::Defn);
+          auto global_name = Match(TokenType::Global).ToString();
+          auto global = GlobalVar(global_name);
+          global_names.Add(global_name, global);
+          auto func = ParseFunctionDef();
+          defs.funcs.push_back(GlobalFunc(global, func));
+          continue;
+        }
+        case TokenType::TypeDef: {
+          defs.types.push_back(ParseTypeDef());
+          continue;
+        }
+        default:
+          return defs;
+      }
+    }
+  }
+
+  TypeData ParseTypeDef() {
+    // Match the `type` keyword.
+    Match(TokenType::TypeDef);
+    // Parse the type's identifier.
+    auto type_id = Match(TokenType::Identifier).ToString();
+    auto type_global = tvm::GlobalTypeVar(type_id, TypeKind::kTypeData);
+    type_names.Add(type_id, type_global);
+
+    Array<TypeVar> generics;
+
+    bool should_pop = false;
+    if (Peek()->token_type == TokenType::LSquare) {
+      // If we have generics we need to add a type scope.
+      PushTypeScope();
+      should_pop = true;
+      generics = ParseSequence<TypeVar>(TokenType::LSquare, TokenType::Comma, TokenType::RSquare, [&]() {
+        auto type_var_name = Match(TokenType::Identifier).ToString();
+        return BindTypeVar(type_var_name, TypeKind::kType);
+      });
+    }
+
+    // Parse the list of constructors.
+    auto ctors = ParseSequence<tvm::Constructor>(TokenType::LCurly, TokenType::Comma, TokenType::RCurly, [&]() {
+      // First match the name of the constructor.
+      auto ctor = Match(TokenType::Identifier).ToString();
+      // Match the optional field list.
+      if (Peek()->token_type != TokenType::OpenParen) {
+        return tvm::Constructor(ctor, {}, type_global);
+      } else {
+        auto arg_types = ParseSequence<Type>(TokenType::OpenParen, TokenType::Comma, TokenType::CloseParen, [&]() {
+          return ParseType();
+        });
+        return tvm::Constructor(ctor, arg_types, type_global);
+      }
+    });
+
+    // Now pop the type scope.
+    if (should_pop) {
+      PopTypeScopes(1);
+    }
+
+    return TypeData(type_global, generics, ctors);
+  }
+
+  template <typename R>
+  R Bracket(TokenType open, TokenType close, std::function<R()> parser) {
+    Match(open);
+    R result = parser();
+    Match(close);
+    return result;
+  }
+
+  template <typename R>
+  R Parens(std::function<R()> parser) {
+    return Bracket(TokenType::OpenParen, TokenType::CloseParen, parser);
+  }
+
+  template <typename R>
+  R Block(std::function<R()> parser) {
```

Comment:
const & for std::function

Reply:
It seems that if you immediately consume you get better behavior with move semantics which is what we are doing here.